File: 2. Structural Design Patterns/Proxy/C#/ProxyExample.cs
// Proxy Pattern Example in C#
// Image viewer with lazy-loading proxy

using System;

namespace ProxyExample
{
    interface IImage { void Display(); }

    class RealImage : IImage
    {
        private readonly string _filename;
        public RealImage(string filename){ _filename = filename; LoadFromDisk(); }
        private void LoadFromDisk(){ Console.WriteLine($"Loading image: {_filename}"); }
        public void Display(){ Console.WriteLine($"Displaying: {_filename}"); }
    }

    class ProxyImage : IImage
    {
        private readonly string _filename;
        private RealImage _real;
        public ProxyImage(string filename){ _filename = filename; }
        public void Display(){ _real ??= new RealImage(_filename); _real.Display(); }
    }

    class Program
    {
        static void Main()
        {
            IImage img = new ProxyImage("photo.jpg");
            img.Display(); // loads then displays
            img.Display(); // displays without loading
        }
    }
}

File: 2. Structural Design Patterns/Proxy/Java/ProxyExample.java
// Proxy Pattern Example in Java
// Image viewer with lazy-loading proxy

interface Image { void display(); }

class RealImage implements Image {
    private final String filename;
    RealImage(String filename){ this.filename = filename; loadFromDisk(); }
    private void loadFromDisk(){ System.out.println("Loading image: " + filename); }
    public void display(){ System.out.println("Displaying: " + filename); }
}

class ProxyImage implements Image {
    private final String filename;
    private RealImage real;
    ProxyImage(String filename){ this.filename = filename; }
    public void display(){ if (real == null) real = new RealImage(filename); real.display(); }
}

public class ProxyExample {
    public static void main(String[] args){ Image img = new ProxyImage("photo.jpg"); img.display(); img.display(); }
}
